src/libsyntax/ptr.rs

```diff
@@ -101,7 +102,9 @@ impl<T: 'static> P<T> {
                 // Recreate self from the raw pointer.
                 Some(P { ptr: Box::from_raw(p) })
             } else {
-                drop(Box::from_raw(p));
+                // Free the allocated memory only,
+                // do not drop T as it was ptr::read above.
+                drop(Box::from_raw(p as *mut ManuallyDrop<T>));
```

Comment:
cc @RalfJung this is a clever trick I didn't consider!

Reply:
Yeah I think this should work. I don't like the raw ptr cast though, it's so easy to get those wrong -- but I also don't know another way here.